cli/command/stack/list.go

```diff
@@ -18,18 +32,168 @@ func newListCommand(dockerCli command.Cli) *cobra.Command {
+			opts.AllNamespaces = true
+		}
+		if opts.AllNamespaces {
+			stacks, err = getStacksWithAllNamespaces(stacks, dockerCli, opts, flags)
+			if err != nil {
+				return err
+			}
+		} else {
+			stacks, err = getStacksWithNamespaces(stacks, dockerCli, opts, flags)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return format(dockerCli, opts, stacks)
+}
+
+func getStacksWithAllNamespaces(stacks []*formatter.Stack, dockerCli command.Cli, opts options.List, flags *pflag.FlagSet) ([]*formatter.Stack, error) {
+	var err error
+	stacks, err = getStacks(stacks, dockerCli, opts, kubernetes.NewOptions(flags))
+	if err == nil || !errors.IsForbidden(err) {
+		return stacks, err
+	}
+	nms, err := getUserVisibleNamespaces(dockerCli, err)
+	if err != nil {
+		return nil, err
+	}
+	opts.AllNamespaces = false
+	for _, nm := range nms.Items {
+		stacks, err = getStacks(stacks, dockerCli, opts, kubernetes.NewOptions(flags, nm.Name))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return stacks, nil
+}
+
+func getStacksWithNamespaces(stacks []*formatter.Stack, dockerCli command.Cli, opts options.List, flags *pflag.FlagSet) ([]*formatter.Stack, error) {
+	mnms, err := getNamespaces(flags)
+	if err != nil {
+		return nil, err
+	}
+	for nm := range mnms {
+		stacks, err = getStacks(stacks, dockerCli, opts, kubernetes.NewOptions(flags, nm))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return stacks, nil
+}
+
+func getStacks(stacks []*formatter.Stack, dockerCli command.Cli, opts options.List, kopts kubernetes.Options) ([]*formatter.Stack, error) {
+	kli, err := kubernetes.WrapCli(dockerCli, kopts)
+	if err != nil {
+		return nil, err
+	}
+	ss, err := kubernetes.GetStacks(kli, opts)
+	if err != nil {
+		return nil, err
+	}
+	return append(stacks, ss...), nil
+}
+
+func getNamespaces(flags *pflag.FlagSet) (map[string]struct{}, error) {
+	nms, err := flags.GetStringSlice("namespace")
+	if err != nil {
+		return nil, err
+	}
+	mnms := map[string]struct{}{}
+	for _, nm := range nms {
+		mnms[nm] = struct{}{}
+	}
+	return mnms, nil
+}
+
+func getUserVisibleNamespaces(dockerCli command.Cli, forbiddenErr error) (*core_v1.NamespaceList, error) {
+	host := dockerCli.Client().DaemonHost()
+	endpoint, err := url.Parse(host)
+	if err != nil || endpoint.Scheme != "tcp" {
+		return nil, forbiddenErr
+	}
+	httpClient := &http.Client{}
+	tlsOptions := dockerCli.ClientInfo().TLSOptions
+	if tlsOptions != nil {
+		tlsConfig, err := tlsconfig.Client(*tlsOptions)
+		if err != nil {
+			return nil, err
+		}
+		httpClient.Transport =
+			&http.Transport{
+				TLSClientConfig: tlsConfig,
+			}
+	}
+	endpoint.Scheme = "https"
+	endpoint.Path = "/kubernetesNamespaces"
+	return listUserNamespaces(httpClient, *endpoint)
+}
+
+func listUserNamespaces(httpClient *http.Client, endpoint url.URL) (*core_v1.NamespaceList, error) {
+	resp, err := httpClient.Get(endpoint.String())
+	if err != nil {
+		return nil, fmt.Errorf("unable to list user namespaces: %s", err)
```

Comment:
Here (and later on), prefer to use `github.com/pkg/errors` with `Wrap` or `Wrapf` : 

```go
return nil, errors.Wrap(err, "unable to list user namespaces")
```